concent_api/middleman/middleman_server.py

```diff
@@ -120,52 +124,86 @@ def _close_middleman(self) -> None:
         self._loop.close()
 
     async def _handle_concent_connection(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        tasks = []
+        response_queue = asyncio.Queue(loop=self._loop)
+        self._connection_id = (self._connection_id + 1) % CONNECTION_COUNTER_LIMIT
+        self._response_queue_pool[self._connection_id] = response_queue
         try:
-            frame = await handle_frame_receive_async(reader, settings.CONCENT_PUBLIC_KEY)
-            remote_address = writer.get_extra_info('peername')
-            print("Received %r from %r" % (frame, remote_address))
-            await self._respond_to_user(frame, writer)
-        except (BrokenEscapingInFrameMiddlemanProtocolError, asyncio.LimitOverrunError, MiddlemanProtocolError) as exception:
-            await self._send_immediate_error(writer, exception)
+            request_producer_task = self._loop.create_task(
+                request_producer(self._request_queue, response_queue, reader, self._connection_id)
+            )
+            response_consumer_task = self._loop.create_task(
+                response_consumer(response_queue, writer, self._connection_id)
+            )
+            tasks.append(request_producer_task)
+            tasks.append(response_consumer_task)
+            await request_producer_task
+            await asyncio.sleep(PROCESSING_TIMEOUT)
+            await response_queue.join()
+            response_consumer_task.cancel()
+
         except Exception as exception:  # pylint: disable=broad-except
             crash_logger.error(
                 f"Exception occurred: {exception}, Traceback: {traceback.format_exc()}"
             )
             raise
+        finally:
+            for task in tasks:
+                task.cancel()
```

Comment:
Successfully completed task should be cancelled too ?

Reply:
there is no harm in cancelling already done task - as it is in `finally` block, we do not know, whether we got here normally, or through `exception`, so it might happen that no task is finished and both need cancelling. I will write comment in the code.
done;